### PR TITLE
refactor: disable insecure serving in controller-manager

### PR DESCRIPTION
**Pull request by knight42**
**What type of PR is this?**

/kind deprecation
/kind cleanup

**What this PR does / why we need it**:

Disable insecure serving in controller-manager

**Which issue(s) this PR fixes**:

xref #91506

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
[ACTION REQUIRED] controller-manager: the following flags have no effect and would be removed in v1.24:
* `--port`
* `--address`
The insecure port flags `--port` may only be set to 0 now.

In addtion, please be careful that:
* controller-manager MUST start with `--authorization-kubeconfig` and `--authentication-kubeconfig` correctly set to get authentication/authorization working.
* liveness/readiness probes to controller-manager MUST use HTTPS now, and the default port has been changed to 10257.
* Applications that fetch metrics from controller-manager should use a dedicated service account which is allowed to access nonResourceURLs `/metrics`.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kube-controller-manager/app/options/options.go [hunk 1]
```diff
@@ -189,7 +183,7 @@ func NewKubeControllerManagerOptions() (*KubeControllerManagerOptions, error) {
 
 	s.Authentication.RemoteKubeConfigFileOptional = true
 	s.Authorization.RemoteKubeConfigFileOptional = true
-	s.Authorization.AlwaysAllowPaths = []string{"/healthz"}
+	s.Authorization.AlwaysAllowPaths = []string{"/healthz", "/metrics"}
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/options/options.go:186]: I wasn't expecting to expose metrics to all callers. What unauthenticated callers are fetching metrics today?

@deads2k @sttts, thoughts here?

**Comment by liggitt** [comment]
Looks like a couple legitimate CI failures, and the one comment about not exposing /metrics. Looks reasonable otherwise.

File: cmd/kube-controller-manager/app/options/options.go (hunk 1 above)
Comment (knight42) [reply cmd/kube-controller-manager/app/options/options.go:186]: > What unauthenticated callers are fetching metrics today?

As far as I know, if a user fetch metrics from controller-manager using prometheus, it is almost certain that the request is unauthenticated.

**Comment by knight42** [comment]
The integration job fails because the test [`TestComponentSecureServingAndAuth`](https://github.com/kubernetes/kubernetes/blob/3f6ec26eefcae1cdb4a43aa4e664a5f60fc7ca78/test/integration/serving/serving_test.go#L78) still accesses the insecure port, looks like we have to split the test cases of controller-manager from the others.

The `e2e-gce-100-performance` job fails because this test would setup a prometheus instance, and the insecure port of controller-manager is hardcoded in the config, so prometheus thinks the controller-manager is not healthy and refuse to handle request:
* https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/96216/pull-kubernetes-e2e-gce-100-performance/1323961326933905408/build-log.txt
```
W1104 12:35:22.446] I1104 12:35:22.446327   97797 prometheus.go:318] Waiting for Prometheus stack to become healthy...
W1104 12:35:52.483] W1104 12:35:52.483504   97797 util.go:64] error while calling prometheus api: the server is currently unable to handle the request (get services http:prometheus-k8s:9090), response: "k8s\x00\n\f\n\x02v1\x12\x06Status\x12g\n\x06\n\x00\x12\x00\x1a\x00\x12\aFailure\x1a=no endpoints available for service \"http:prometheus-k8s:9090\"\"\x12ServiceUnavailable0\xf7\x03\x1a\x00\"\x00"
W1104 12:36:22.661] I1104 12:36:22.661126   97797 util.go:93] 102/115 targets are ready, example not ready target: {map[endpoint:http instance:10.64.15.2:8080 job:prometheus-operator namespace:monitoring pod:prometheus-operator-d7f7f4fcb-cl5sm service:prometheus-operator] unknown}
W1104 12:36:52.552] I1104 12:36:52.552396   97797 util.go:93] 114/115 targets are ready, example not ready target: {map[endpoint:kube-controller-manager instance:10.40.0.3:10252 job:master namespace:monitoring service:master] down}
W1104 12:37:22.611] I1104 12:37:22.610761   97797 util.go:93] 114/115 targets are ready, example not ready target: {map[endpoint:kube-controller-manager instance:10.40.0.3:10252 job:master namespace:monitoring service:master] down}
```
* related manifest: https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/prometheus/manifests/master-ip/master-endpoints.yaml

**Comment by knight42** [comment]
The `pull-kubernetes-e2e-gce-100-performance` job is a bit tricky to fix.. The Prometheus setup in this job would only grab metrics from the insecure port of the controller-manager, if we disable the insecure port of controller-manager, this job is expected to fail.

I think we should update the Prometheus config used in `perf-tests`: https://github.com/kubernetes/perf-tests/blob/9e36dd083b3e4c2bb7ab507ca7b18ee2fdb3d2b2/clusterloader2/pkg/prometheus/manifests/master-ip/master-serviceMonitor.yaml#L43-L44 first?

@liggitt Any thoughts?

**Comment by liggitt** [comment]
> @liggitt Any thoughts?

I'd defer to @sttts on this one... how are you running this e2e with a secured controller manager that disables the unsecured port?

File: cmd/kube-controller-manager/app/options/options.go (hunk 1 above)
Comment (paulfantom) [reply cmd/kube-controller-manager/app/options/options.go:186]: Prometheus itself is capable of fetching metrics from endpoints that require authentication. Currently, basic auth and bearer tokens are supported methods as can be seen in [`scrape_config` section of prometheus configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config). In case of kubernetes, a config like `bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token` is used frequently to authenticate to endpoints. This behavior can be seen in one of the examples: https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus-kubernetes.yml#L59-L87

Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:186]: Afaik in OpenShift Prometheus fetches metrics with a service account aka a token via delegated authentication.

File: cmd/kube-controller-manager/app/options/options.go [hunk 2]
```diff
@@ -209,18 +203,32 @@ func NewKubeControllerManagerOptions() (*KubeControllerManagerOptions, error) {
...
+		bindAddr = net.IPv4(127, 0, 0, 1)
+		bindPort = 0
+	)
+	fs.IPVar(&bindAddr, "address", bindAddr,
+		"The IP address on which to serve the insecure --port (set to 0.0.0.0 for all IPv4 interfaces and :: for all IPv6 interfaces).")
+	fs.MarkDeprecated("address", "This flag has no effect now and will be removed in v1.24.")
+
+	fs.IntVar(&bindPort, "port", bindPort, "The port on which to serve unsecured, unauthenticated access. Set to 0 to disable.")
+	fs.MarkDeprecated("port", "This flag has no effect now and will be removed in v1.24.")
+}
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/options.go:230]: we keep the flags but remove the logic? How strange.

File: cmd/kube-controller-manager/app/options/options_test.go [hunk 3]
```diff
@@ -170,8 +168,8 @@ func TestAddFlags(t *testing.T) {
 	expected := &KubeControllerManagerOptions{
 		Generic: &cmoptions.GenericControllerManagerConfigurationOptions{
 			GenericControllerManagerConfiguration: &cmconfig.GenericControllerManagerConfiguration{
-				Port:            10252,     // Note: InsecureServingOptions.ApplyTo will write the flag value back into the component config
-				Address:         "0.0.0.0", // Note: InsecureServingOptions.ApplyTo will write the flag value back into the component config
+				Port:            10257,     // Note: SecureServingOptions.ApplyTo will write the flag value back into the component config
+				Address:         "0.0.0.0", // Note: SecureServingOptions.ApplyTo will write the flag value back into the component config
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/options_test.go:172]: I expected these to be pruned everywhere. They are old i.e. insecure only, aren't they?

File: cmd/kube-controller-manager/app/options/options.go (hunk 2 above)
Comment (liggitt) [reply cmd/kube-controller-manager/app/options/options.go:230]: We keep the flags that people have been using for years to turn off insecure serving and allow them to continue setting `--port=0` for a few releases. Any other value causes an error.

The reason is to avoid accidents in multi-version deployers (like kubeadm) which could lead to dropping the `--port` flag in 1.20 because it was no longer allowed, but then accidentally re-enabling insecure serving when using those manifests to deploy a 1.19 cluster

Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:230]: got it

File: cmd/kube-controller-manager/app/options/options.go (hunk 1 above)
Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:186]: I would also consider this a regression, even security relevant: there are many parties disabling insecure ports and expecting delegated authn for the secure port. This suddenly exposes metrics to everybody. We shouldn't do this.

Comment (knight42) [reply cmd/kube-controller-manager/app/options/options.go:186]: Got it, thanks for pointing it out.

File: cmd/kube-controller-manager/app/options/options_test.go (hunk 3 above)
Comment (knight42) [reply cmd/kube-controller-manager/app/options/options_test.go:172]: okay I will remove these.

**Comment by cici37** [comment]
Suspect this PR might be affected by: kubernetes/test-infra#19910

**Comment by knight42** [comment]
@cici37 Ah exactly! this problem is really annoying..

**Comment by knight42** [comment]
@liggitt @sttts Put aside the job `pull-kubernetes-e2e-gce-100-performance`, I have refactored the e2e test a bit in this commit 6b718e4c96cfbc3285e37be05ebf030ed5f3ed09 to make the job `pull-kubernetes-e2e-gce-ubuntu-containerd` succeed. 

In summary, my primary changes include:
1. pass flags `--authentication-kubeconfig` and `--authorization-kubeconfig` to controller-manager
2. setup RBAC rules to allow anonymous access to `/metrics` endpoint in the test

How are these changes sound to you?

As for the job `pull-kubernetes-e2e-gce-100-performance`, I have no instant solution now and need some time to investigate.

File: cmd/kube-controller-manager/app/options/options_test.go [hunk 4]
```diff
@@ -171,8 +169,7 @@ func TestAddFlags(t *testing.T) {
 	expected := &KubeControllerManagerOptions{
 		Generic: &cmoptions.GenericControllerManagerConfigurationOptions{
 			GenericControllerManagerConfiguration: &cmconfig.GenericControllerManagerConfiguration{
-				Port:            10252,     // Note: InsecureServingOptions.ApplyTo will write the flag value back into the component config
-				Address:         "0.0.0.0", // Note: InsecureServingOptions.ApplyTo will write the flag value back into the component config
+				Address:         "0.0.0.0", // Note: This field should have no effect in CM now, and "0.0.0.0" is the default value.
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/options/options_test.go:172]: @sttts - I expected these values to be defaulted and effective. If this has no effect, what field controls the listen address/port?

File: cmd/kube-controller-manager/app/options/options_test.go [hunk 5]
```diff
@@ -461,8 +453,6 @@ func TestApplyTo(t *testing.T) {
 	expected := &kubecontrollerconfig.Config{
 		ComponentConfig: kubectrlmgrconfig.KubeControllerManagerConfiguration{
 			Generic: cmconfig.GenericControllerManagerConfiguration{
-				Port:            10252,     // Note: InsecureServingOptions.ApplyTo will write the flag value back into the component config
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/options/options_test.go:464]: same question for these... where is the listen address/port set if not here?

File: staging/src/k8s.io/controller-manager/options/generic.go [hunk 6]
```diff
@@ -75,8 +75,6 @@ func (o *GenericControllerManagerConfigurationOptions) ApplyTo(cfg *cmconfig.Gen
 		return err
 	}
 
-	cfg.Port = o.Port
-	cfg.Address = o.Address
```
Comment (liggitt) [inline staging/src/k8s.io/controller-manager/options/generic.go:79]: why are we not plumbing these?

File: cmd/kube-controller-manager/app/options/options_test.go (hunk 4 above)
Comment (knight42) [reply cmd/kube-controller-manager/app/options/options_test.go:172]: IIUC these settings seem to be used for the insecure serving, I'd like @sttts to confirm.

> If this has no effect, what field controls the listen address/port?
Flag `--bind-address` and `--secure-port`:
* https://github.com/kubernetes/kubernetes/blob/75463bda1d9103a46de127528ee7b3b96ceada54/cmd/kube-controller-manager/app/options/options.go#L235
* https://github.com/kubernetes/kubernetes/blob/75463bda1d9103a46de127528ee7b3b96ceada54/cmd/kube-controller-manager/app/options/options.go#L357-L367
* https://github.com/kubernetes/kubernetes/blob/75463bda1d9103a46de127528ee7b3b96ceada54/cmd/kube-controller-manager/app/controllermanager.go#L193-L200

File: cmd/kube-controller-manager/app/options/options_test.go (hunk 5 above)
Comment (knight42) [reply cmd/kube-controller-manager/app/options/options_test.go:464]: answered above https://github.com/kubernetes/kubernetes/pull/96216#discussion_r522169557

File: staging/src/k8s.io/controller-manager/options/generic.go (hunk 6 above)
Comment (knight42) [reply staging/src/k8s.io/controller-manager/options/generic.go]: After taking a second look, I found out that `cloud-provider` would call this function https://github.com/kubernetes/kubernetes/blob/0e0cc1ead8ae5937a45e8ca9fa68eb300a1a2f50/staging/src/k8s.io/cloud-provider/options/options.go#L152, so I have reverted this change. Thanks for pointing it out.

**Comment by knight42** [comment]
@liggitt Hi, now this is being blocked by `clusterloader2` again, it seems to be a dependency problem this time.

According to the log https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/96216/pull-kubernetes-e2e-gce-100-performance/1348636918799142912/
```
W0111 15:03:22.035] E0111 15:03:22.035028  124783 metrics_for_e2e.go:79] MetricsForE2E: metricsGrabber failed to grab some of the metrics: Errors while grabbing metrics: [error waiting for controller manager pod to expose metrics: timed out waiting for the condition; an error on the server ("unknown") has prevented the request from succeeding (get pods kube-controller-manager-e2e-96216-95a39-master:10252)]
```

`clusterloader2` fails to grab metrics from controller-manager because we disable the insecure port in this PR, and it turns out that `clusterloader2` actually ultilizes the e2e framework in k/k repo to fetch metrics:
* https://github.com/kubernetes/perf-tests/blob/16a8cfc62b991869cad750ea07d9d74332f76af0/clusterloader2/pkg/measurement/common/metrics_for_e2e.go#L65
* https://github.com/kubernetes/perf-tests/blob/16a8cfc62b991869cad750ea07d9d74332f76af0/clusterloader2/pkg/measurement/common/metrics_for_e2e.go#L23

We seem to stuck in a dead-lock again..

@liggitt I would like to hear your advice. Should we refactor `clusterloader2` to make it fetch metrics without using `k8s.io/kubernetes/test/e2e/framework/metrics` first? Or could we sidestep this issue?

**Comment by liggitt** [comment]
cc @mm4tt @kubernetes/sig-scalability on how to get scale tests the results they need when the controller manager insecure port is disabled

**Comment by wojtek-t** [comment]
> cc @mm4tt @kubernetes/sig-scalability on how to get scale tests the results they need when the controller manager insecure port is disabled

I don't see a fast and easy workaround.
I think the fastest thing we can do would be to fork the metricsGrabber code into perf-tests repo and fix the issue there (ugly but would allow to unblock you) [I would ideally stop vendoring test/e2e from k/k code in perf-tests anyway, so that's step towards this direction.]

**Comment by knight42** [comment]
> I think the fastest thing we can do would be to fork the metricsGrabber code into perf-tests repo and fix the issue there

@wojtek-t  Okay... I will submit a PR tomorrorw.

FWIW below is the result I ran `ripgrep -g '!vendor/*' -g '!go.*' 'k8s.io/kubernetes'` under `clusterloader2` repo:
```
pkg/provider/vsphere.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/provider/gke.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/provider/skeleton.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/provider/local.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/provider/gce.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/provider/util.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/provider/aws.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/provider/kubemark.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/provider/eks.go:	sshutil "k8s.io/kubernetes/pkg/ssh"
pkg/measurement/util/kubelet/kubelet.go:	stats "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
pkg/measurement/util/runtimeobjects/runtimeobjects.go:	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
pkg/measurement/common/scheduler_latency.go:	"k8s.io/kubernetes/pkg/master/ports"
pkg/measurement/common/scheduler_latency.go:	schedulermetric "k8s.io/kubernetes/pkg/scheduler/metrics"
pkg/measurement/common/metrics_for_e2e.go:	"k8s.io/kubernetes/test/e2e/framework/metrics"
cmd/clusterloader.go:	"k8s.io/kubernetes/pkg/master/ports"
```

**Comment by knight42** [comment]
@liggitt Finally all tests passed now. Please take another look.

File: staging/src/k8s.io/controller-manager/options/generic.go (hunk 6 above)
Comment (liggitt) [reply staging/src/k8s.io/controller-manager/options/generic.go:79]: this still looks wrong... did you forget to push?

File: cmd/kube-controller-manager/app/options/options_test.go (hunk 4 above)
Comment (liggitt) [reply cmd/kube-controller-manager/app/options/options_test.go:172]: if port and address are unused, should they be removed? I'm surprised the insecure serving config wasn't represented in the config types

File: staging/src/k8s.io/controller-manager/options/generic.go (hunk 6 above)
Comment (knight42) [reply staging/src/k8s.io/controller-manager/options/generic.go]: nice catch.. fixed.

**Comment by liggitt** [comment]
This needs a much more detailed release note, since it means that deployments may have to change their kube-controller-manager manifests and monitoring.

* Manifest liveness/readiness probes must be https now, and the default port is different
* Things monitoring controller-manager metrics must be authenticated+authorized (or anonymous access to /metrics must be granted... which seems bad, since it is not scoped to the controller-manager and would also give anonymous access to apiserver+scheduler metrics, etc)
* Is it even possible for things accessing controller-manager metrics via the kube-apiserver proxy to authenticate? They can't propagate a bearer token or a client certificate

File: test/e2e/e2e.go [hunk 7]
```diff
@@ -302,6 +303,24 @@ func setupSuite() {
 		nodeKiller := framework.NewNodeKiller(framework.TestContext.NodeKiller, c, framework.TestContext.Provider)
 		go nodeKiller.Run(framework.TestContext.NodeKiller.NodeKillerStopCh)
 	}
+
+	// Allow anonymous access to `/metrics` in tests
```
Comment (liggitt) [inline test/e2e/e2e.go:307]: hmm... is this what we expect anyone accessing kcm metrics via the kube-apiserver proxy to do? this also grants access to apiserver /metrics to all anonymous users

Comment (knight42) [reply test/e2e/e2e.go:307]: >  is this what we expect anyone accessing kcm metrics via the kube-apiserver proxy to do?

Currenly I could not think of a better solution. My original purpose of this RBAC rule is to allow anonymous access to `/metrics` **ONLY** in tests, since it is a bit hard to get tests passed after turning off the insecure port. I figured out this way by learning from `perf-tests`: https://github.com/kubernetes/perf-tests/blob/ca470d00ba5e62f77b8afefc8bcb4d5440420fd4/clusterloader2/pkg/prometheus/prometheus.go#L244

Comment (knight42) [reply test/e2e/e2e.go:307]: I think the point is how sensitive the metrics is. 
If metrics is not sensitive, then it is okay to be allow public access to metrics. 
If metrics is sensitive, is it still legitimate to access kcm metrics via the kube-apiserver proxy(which drops Authorization header and doesn't expect pods to see the credential)? Should we use a dedicated service account here?

**Comment by knight42** [comment]
@liggitt I have updated the release note.

File: test/e2e/e2e.go (hunk 7 above)
Comment (liggitt) [reply test/e2e/e2e.go:307]: @deads2k @sttts curious how you get metrics from the controller-manager and scheduler pods. any input here?

Comment (sttts) [reply test/e2e/e2e.go:307]: We do token reviews and authn via normal RBAC.

Comment (liggitt) [reply test/e2e/e2e.go:307]: > We do token reviews and authn via normal RBAC.

so you access the pod(s) directly or via service, not via the apiserver proxy

Comment (sttts) [reply test/e2e/e2e.go:307]: @s-urbaniak via service endpoints?

Comment (s-urbaniak) [reply test/e2e/e2e.go:307]: hello :wave: :-) for kcm, we usually pick up metrics via services and the following selectors: https://github.com/prometheus-operator/kube-prometheus/blob/1bf43811174355359e5316b52bfb1a0b928550b2/manifests/prometheus-serviceMonitorKubeControllerManager.yaml#L54-L59

Comment (liggitt) [reply test/e2e/e2e.go:307]: I don't think exposing all /metrics on all components to all anonymous users is reasonable... if we have to do it for this e2e test to get scheduler/kube-controller-manager metrics, it is likely downstream consumers will have to do the same. I'd like to see this switch to pulling metrics from the pod in an authenticated way (maybe via a command run in a collector pod or something? not sure...) that we can point to as an example of how to do it properly.

Comment (knight42) [reply test/e2e/e2e.go]: @liggitt How about we add a command to `agnhost`, say `proxy-metrics`, which would listen on `10257` and `10259` , and grab metrics from the secure port of controller-manager and scheduler respectively using the service account token when users access these ports, then we could use it as a reverse-proxy in tests.

Comment (liggitt) [reply test/e2e/e2e.go]: since anyone actually using the apiserver proxy to fetch metrics from the controlplane pods would have to make adjustments as well, I think we need an approach we could recommend to production consumers as well, which probably means it should not be in a test-oriented pod?

Comment (knight42) [reply test/e2e/e2e.go]: @liggitt All tests passed now. I decided to setup a nginx to expose the metrics in the test.

**Review by liggitt — COMMENTED** [review]
one question about the port/address assignments.

would like @sttts to review the approach used to access the controller manager over the secure port.

/assign @sttts

File: cmd/kube-controller-manager/app/options/options.go [hunk 8]
```diff
@@ -365,12 +369,6 @@ func (s *KubeControllerManagerOptions) ApplyTo(c *kubecontrollerconfig.Config) e
 			return err
 		}
 	}
-
-	// sync back to component config
-	// TODO: find more elegant way than syncing back the values.
-	c.ComponentConfig.Generic.Port = int32(s.InsecureServing.BindPort)
-	c.ComponentConfig.Generic.Address = s.InsecureServing.BindAddress.String()
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:372]: @knight42 have you checked how those fields are used?

[the remark this answers is not in the document]
Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:372]: @ingvagabund is looking into the same for kube-scheduler. Hopefully he has an overview about this mess.

[the remark this answers is not in the document]
Comment (ingvagabund) [reply cmd/kube-controller-manager/app/options/options.go:372]: From https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/controller-manager/config/types.go#L27-L28:
```
// GenericControllerManagerConfiguration holds configuration for a generic controller-manager
type GenericControllerManagerConfiguration struct {
	// port is the port that the controller-manager's http service runs on.
	Port int32
```

The comment says http service, not https. Also, `.Generic.Port` does not look to be used anywhere else in KCM.

[the remark this answers is not in the document]
Comment (ingvagabund) [reply cmd/kube-controller-manager/app/options/options.go:372]: No reason to set any of port/address fields anymore.

[the remark this answers is not in the document]
Comment (ingvagabund) [reply cmd/kube-controller-manager/app/options/options.go:372]: Both secure-port and bind address flags are set through k8s.io/apiserver: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/serving.go#L163. Which are then put into KCM through SecureServingOptions (NewSecureServingOptions()) in https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/options/options.go#L182.

Making Port and Address from the component config obsolete.

[the remark this answers is not in the document]
Comment (knight42) [reply cmd/kube-controller-manager/app/options/options.go:371]: @sttts Those fields are only used for insecure serving as I described in https://github.com/kubernetes/kubernetes/pull/96216#discussion_r522169557 and @ingvagabund 's comment also confirms that.

[the remark this answers is not in the document]
Comment (knight42) [reply cmd/kube-controller-manager/app/options/options.go:371]: It would be great to see this PR got landed before the upcoming code freeze 😄

File: cmd/kube-controller-manager/app/options/options.go [hunk 9]
```diff
@@ -212,18 +206,31 @@ func NewKubeControllerManagerOptions() (*KubeControllerManagerOptions, error) {
 }
 
 // NewDefaultComponentConfig returns kube-controller manager configuration object.
-func NewDefaultComponentConfig(insecurePort int32) (kubectrlmgrconfig.KubeControllerManagerConfiguration, error) {
+func NewDefaultComponentConfig(securePort int32) (kubectrlmgrconfig.KubeControllerManagerConfiguration, error) {
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/options.go:209]: remove securePort, it's not used

File: cmd/kube-controller-manager/app/options/options.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:371]: Have confirmed that they are not used for anything but insecure serving.

File: cmd/kube-controller-manager/app/options/options_test.go (hunk 4 above)
Comment (sttts) [reply cmd/kube-controller-manager/app/options/options_test.go:172]: Yes, would like to see them removed from `GenericControllerManagerConfiguration`. Today they are part of the v1alpha1 config type. Is that used anywhere? Compare k8s.io/controller-manager/config/v1alpha1/types.go

File: test/integration/serving/serving_test.go [hunk 10]
```diff
@@ -237,9 +249,9 @@ func testComponent(t *testing.T, tester componentTester, kubeconfig, brokenKubec
 		}, "/metrics", false, false, intPtr(http.StatusInternalServerError), intPtr(http.StatusOK)},
 		{"always-allowed /metrics with BROKEN authn/authz", []string{
 			"--port=0",
-			"--authentication-skip-lookup", // to survive unaccessible extensions-apiserver-authentication configmap
-			"--authentication-kubeconfig", kubeconfig,
-			"--authorization-kubeconfig", kubeconfig,
+			"--authentication-skip-lookup", // to survive inaccessible extensions-apiserver-authentication configmap
+			"--authentication-kubeconfig", brokenKubeconfig,
+			"--authorization-kubeconfig", brokenKubeconfig,
```
Comment (sttts) [inline test/integration/serving/serving_test.go:254]: These are separate fixes. Please move to their own commit.

File: test/integration/serving/serving_test.go [hunk 11]
```diff
@@ -322,6 +334,111 @@ func testComponent(t *testing.T, tester componentTester, kubeconfig, brokenKubec
...
+				}
+				defer r.Body.Close()
+				if got, expected := r.StatusCode, *tt.wantSecureCode; got != expected {
+					t.Fatalf("expected http %d at %s of component, got: %d %q", expected, tt.path, got, string(body))
+				}
+			}
+		})
+	}
+}
+
```
Comment (sttts) [inline test/integration/serving/serving_test.go:441]: This is lots of duplicated code. How does the variant looks like that works for both cases?

Comment (knight42) [reply test/integration/serving/serving_test.go:435]: This PR only disables insecure serving in `kube-controller-manager`, `kube-scheduler` and `cloud-controller-manager` still have insecure serving enabled. 

I don't want to changge the tests that validates the insecure serving, so I wrote another function `testComponentWithSecureServing` that only validates the secure serving.

**Comment by knight42** [comment]
@sttts Hi, I think I have addressed your comments, please take another look.

**Comment by knight42** [comment]
(randomly picking approvers, please feel free to assign others if I picked the wrong person):
* for `cluster/gce/manifests`: @MrHohn 
* for `pkg`: @thockin 

/assign @MrHohn @thockin

File: pkg/cluster/ports/ports.go [hunk 12]
```diff
@@ -25,10 +25,6 @@ const (
 	// KubeletPort is the default port for the kubelet server on each host machine.
 	// May be overridden by a flag at startup.
 	KubeletPort = 10250
-	// InsecureKubeControllerManagerPort is the default port for the controller manager status server.
-	// May be overridden by a flag at startup.
-	// Deprecated: use the secure KubeControllerManagerPort instead.
-	InsecureKubeControllerManagerPort = 10252
```
Comment (lavalamp) [inline pkg/cluster/ports/ports.go:31]: We don't want this to be reused, so instead of removing it, change the documentation to state what it used to be fore and that it's been permanently turned off.

**Comment by lavalamp** [comment]
I think you can't turn off the insecure ports in the same release that you make communication go over the secure ports, some rollback scenarios are broken if we do that.

File: pkg/registry/core/rest/storage_core.go [hunk 13]
```diff
@@ -343,7 +344,7 @@ func (s componentStatusStorage) serversToValidate() map[string]*componentstatus.
 	// this is fragile, which assumes that the default port is being used
 	// TODO: switch to secure port until these components remove the ability to serve insecurely.
 	serversToValidate := map[string]*componentstatus.Server{
-		"controller-manager": {Addr: "127.0.0.1", Port: ports.InsecureKubeControllerManagerPort, Path: "/healthz"},
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/registry/core/rest/storage_core.go:346]: Per https://kubernetes.io/docs/setup/release/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager:

> kube-controller-manager, kube-scheduler, and cloud-controller-manager must not be newer than the kube-apiserver instances they communicate with.

This also only works for deployments where the controller-manager is colocated with the API server (since it uses localhost communication). In all the deployments I'm aware of, the API server and controller manager move versions together as static pods in lockstep.
